### ISSUE TITLE
Segmentation fault when scanning for projects in the Project Manager

### ISSUE THREAD

**Issue by NilsIrl** [godotengine/godot#29192]
**Godot version:**

0d8f1ba6a9182d1f1af39097e2165fc9807377f5
05447a5
v3.1.1.stable.official

In other words this seems to affect most versions.

**OS/device including version:**

Linux 4.15.0-15-generic Ubuntu 18.04

16GB of RAM (my RAM doesn't get filled)

**Issue description:**

When scanning by home directory (enormous (823G)), in the Project Manager, godot crashes with a Segmentation fault.

This doesn't happen when I scan my Documents folder which is a lot smaller (320G).

**Steps to reproduce:**
1. Open Godot
2. Click on Scan
3. Choose a big directory

This isn't (at least it doesn't seem to be) related to files which needs higher privileges to be read as I copied some into my Documents directory to test.

**Comment by NilsIrl** [comment]
The function which scans for file is the following:
https://github.com/godotengine/godot/blob/0d8f1ba6a9182d1f1af39097e2165fc9807377f5/editor/project_manager.cpp#L1656

**Comment by KoBeWi** [comment]
So I just started scanning my whole C drive to test it (142GB). Godot's been scanning it for like 10 minutes already, even though it's SSD. But what's actually bad is that Godot takes increasingly more RAM with each second, sitting on over 2GB right now. This might be caused by recursive scan while keeping sub-directory list in a list of strings. The segfualt could have happened due to Godot running out of memory.

I wonder if scanning the directory on spot instead of adding it to list would make things any better.

**Comment by NilsIrl** [comment]
I don't know if this is an error from my side but in GDB using the backtrace command, the function _scan_dir is called 58131 times. That is a recursion of depth 58131 which is quite big.

**Comment by NilsIrl** [comment]
Also shouldn't https://github.com/godotengine/godot/blob/0d8f1ba6a9182d1f1af39097e2165fc9807377f5/editor/project_manager.cpp#L1649 be moved out of the for loop?

I don't know how compilers optimize exactly but it's still better to avoid recalculating a value for no reason.

**Comment by NilsIrl** [comment]
> So I just started scanning my whole C drive to test it (142GB). Godot's been scanning it for like 10 minutes already, even though it's SSD. But what's actually bad is that Godot takes increasingly more RAM with each second, sitting on over 2GB right now. This might be caused by recursive scan while keeping sub-directory list in a list of strings. The segfualt could have happened due to Godot running out of memory.

In my case, the Segmentation fault happens after just 10 seconds or so and my RAM definitively doesn't get filled.

**Comment by NilsIrl** [comment]
https://github.com/godotengine/godot/blob/0d8f1ba6a9182d1f1af39097e2165fc9807377f5/editor/project_manager.cpp#L1647
This is the line that segfaults.

```
#0  0x00007ffff50c630e in _int_malloc (av=av@entry=0x7ffff541dc40 <main_arena>, bytes=bytes@entry=33) at malloc.c:3557
#1  0x00007ffff50c78f5 in _int_realloc (av=av@entry=0x7ffff541dc40 <main_arena>, oldp=oldp@entry=0x57ed400, oldsize=oldsize@entry=32, nb=nb@entry=48) at malloc.c:4570
#2  0x00007ffff50caf9b in __GI___libc_realloc (oldmem=0x57ed410, bytes=32) at malloc.c:3230
#3  0x00000000028ccdae in Memory::realloc_static () at core/os/memory.cpp:140
#4  0x000000000045233d in CowData<String>::resize () at ./core/cowdata.h:282
#5  0x0000000000452456 in Vector<String>::resize () at ./core/vector.h:84
#6  Vector<String>::push_back () at ./core/vector.h:152
#7  0x00000000028117cd in String::split () at core/ustring.cpp:788
#8  0x0000000002813406 in String::simplify_path () at core/ustring.cpp:3164
#9  0x0000000000e864f2 in DirAccessUnix::change_dir () at drivers/unix/dir_access_unix.cpp:320
#10 0x000000000110dd95 in ProjectManager::_scan_dir () at editor/project_manager.cpp:1647
#11 0x000000000110dde8 in ProjectManager::_scan_dir () at editor/project_manager.cpp:1650
#12 0x000000000110dde8 in ProjectManager::_scan_dir () at editor/project_manager.cpp:1650
#13 0x000000000110dde8 in ProjectManager::_scan_dir () at editor/project_manager.cpp:1650
```

**Comment by akien-mga** [comment]
Fixed by #29198.
